kernel/src/schema.rs

```diff
@@ -105,6 +106,30 @@ impl StructField {
         self.metadata.get(key.as_ref())
     }
 
+    /// Get the physical name for this field as it should be read from parquet, based on the
+    /// specified column mapping mode.
+    pub fn physical_name(&self, mapping_mode: &ColumnMappingMode) -> DeltaResult<&str> {
+        let name_mapped_name = self
+            .metadata
+            .get(ColumnMetadataKey::ColumnMappingPhysicalName.as_ref());
+        match (mapping_mode, name_mapped_name) {
+            (ColumnMappingMode::None, _) => Ok(self.name.as_str()),
+            (ColumnMappingMode::Name, None) => {
+                Err(Error::MissingData(
+                    format!("No {} key found in field metadata",
+                            ColumnMetadataKey::ColumnMappingPhysicalName.as_ref())
+                ))
+            }
+            (ColumnMappingMode::Name, Some(val)) => match val {
+                MetadataValue::Number(_) => {
+                    Err(Error::generic("{ColumnMetadataKey::ColumnMappingPhysicalName} must be a string in name mapping mode"))
+                }
+                MetadataValue::String(name) => Ok(name),
+            }
+            (ColumnMappingMode::Id, _) => Err(Error::generic("Don't support id column mapping yet")),
```

Comment:
We should think about how to eventually support id mapping mode, because it's likely to be more invasive than name mapping, and name mapping is expressible in terms of id mapping. Maybe we can get away with a single implementation?

Reply:
Actually, we can't get away with a single implementation because (a) field ids are really a problem for the parquet reader to deal with; and (b) existing tables converted to column mapping mode have parquet files whose schema lacks field ids (which is one reason why the parquet reader has to handle field ids). 

But by the same token... if the parquet reader is handling field ids, then maybe id mapping mode isn't so invasive after all. Could kernel "implement" id mapping mode by just verifying, at snapshot creation time, that the table schema contains field ids iff the mode is enabled, so the parquet reader doesn't get confused?

Reply:
Yeah. My feeling was that id mapping will require some changes to either the api, or at least the api contract, that we expose for parquet reading.

regardless, i've heard name mapping is more common and important to support, so merging this and then looking at adding id mapping seems to make sense to me. hopefully it won't require big changes to any of this code, and will mostly just be adding new logic in the places it matters.